Command to get individual keys

NotLebedev opened the pull request:
# Description
Add a `keybindings get` command to listen and get individual "keyboard" events. This includes different keyboard keys (see example of use) on seemingly all terminals and mouse, resize, focus and paste events on some special once. The record returned by this command is similar to crossterm event structure and is documented in help message. For ease of use, option `--types` can get a list of event types to filter only desired events automatically. Additionally `--raw` options displays raw code of char keys and numeric format of modifier flags.

Example of use, moving a character around a grid with arrow keys:
```nu
def test [] {
  mut x = 0
  mut y = 0
  loop {
    clear
    $x = ([([$x 4] | math min) 0] | math max)
    $y = ([([$y 4] | math min) 0] | math max)

    for i in 0..4 {
      for j in 0..4 {
        if $j == $x and $i == $y {
          print -n "*"
        } else {
          print -n "."
        }
      }
      print ""
    }
    
    let inp = (input listen-t [ key ])
    match $inp.key {
      {type: other key: enter} => (break)
      {type: other key: up} => ($y = $y - 1)
      {type: other key: down} => ($y = $y + 1)
      {type: other key: left} => ($x = $x - 1)
      {type: other key: right} => ($x = $x + 1)
      _ => ()
    }
  }
}

```

# User-Facing Changes
- New `keybindngs get` command
- `keybindings listen` is left as is
- New `input display` command in std, mirroring functionality of `keybindings listen`

# Tests + Formatting
<!--
Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect -A clippy::result_large_err` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass
- `cargo run -- crates/nu-std/tests/run.nu` to run the tests for the standard library

> **Note**
> from `nushell` you can also use the `toolkit` as follows
> ```bash
> use toolkit.nu  # or use an `env_change` hook to activate it automatically
> toolkit check pr
> ```
-->

# After Submitting
<!-- If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date. -->

sholderbach commented:
That is a great idea!
From a consistency perspective maybe we can move this to the `input` command as a subcommand? In theory nothing would be stopping us from then reimplementing `keybindings listen` just in terms of your new command.

As a caveat I think we will have to give the general structure of how the keybindings are structured a hard look before commiting to it with 1.0 as there is still some weird reedline dev-history embedded into it. But this shouldn't affect the pure key information too much. only what additional metadata we pull from `crossterm` may change a bit (e.g. the key up/repeat/down state that is only available with some terminals)

NotLebedev commented:
> From a consistency perspective maybe we can move this to the input command as a subcommand? In theory nothing would be stopping us from then reimplementing keybindings listen just in terms of your new command.

Indeed. Just need some good name. The caveat is this is really not only about keybindings, but also mouse, focus resizing. 

> only what additional metadata we pull from crossterm may change a bit (e.g. the key up/repeat/down state that is only available with some terminals)

Actually I already did some digging here. Currently windows-specific (as far as I can tell) separate up-down events are suppressed. There are more of this things specific to windows only, kitty terminal only e.t.c. Currently I did not really investigate this too deep, only the most surface-level things and did not really document it. Perhaps it would be a good idea to work out some subset of events that this command should report and define it strictly. I think behavior across OS and terminals must be as similar as possible, even if somewhat limited

fdncred commented on crates/nu-cli/src/commands/keybindings_get.rs:
```diff
@@ -0,0 +1,354 @@
+    }
+
+    fn usage(&self) -> &str {
+        "Get keyboard event from user"
```
I'm thinking `input listen` as a command name? It seems like it's what this does and it's not limited to keyboard events.
```suggestion
        "Listen for user interface event"
```

NotLebedev replied:
Done

fdncred commented:
Seems very cool to me. Added my vote for command name above. Gonna play with this for a while now. :)

fdncred commented:
unrelated to the pr, but in the script, to get rid of the flicker, i wonder if we could use ansi escapes to draw this grid. it's just an idea to play around with. i bet we could use A B C D from the table below.

ansi --list shows
ansi cursor_off and ansi cursor_on



ESC Code Sequence | Description
-- | --
ESC[H | moves cursor to home position (0, 0)
ESC[{line};{column}HESC[{line};{column}f | moves cursor to line #, column #
ESC[#A | moves cursor up # lines
ESC[#B | moves cursor down # lines
ESC[#C | moves cursor right # columns
ESC[#D | moves cursor left # columns
ESC[#E | moves cursor to beginning of next line, # lines down
ESC[#F | moves cursor to beginning of previous line, # lines up
ESC[#G | moves cursor to column #
ESC[6n | request cursor position (reports as ESC[#;#R)
ESC M | moves cursor one line up, scrolling if needed
ESC 7 | save cursor position (DEC)
ESC 8 | restores the cursor to the last saved position (DEC)
ESC[s | save cursor position (SCO)
ESC[u | restores the cursor to the last saved position (SCO)

fdncred commented:
I'm thinking through this command, testing it and I like it but....
1. I'm not sure I like returning a record vs a table. I think a record is helpful for automation, like your script, but a table is more user friendly.
For instance, I do this and it's not super helpful
```
❯ keybindings get
╭───────────┬───────────────────╮
│ type      │ key               │
│ key       │ {record 2 fields} │
│ modifiers │ [list 0 items]    │
╰───────────┴───────────────────╯
```
This is actually really annoying because I always forget to add `| table -e` so I have to run this command twice to see what I wanted.


2. I'd like to see some examples of each type of capture. I'm not sure how to do anything other than keys.
For example, I was testing the paste functionality and got this
![image](https://github.com/nushell/nushell/assets/343840/d10ab3b5-d07c-4d9f-96bb-6b1bd50f0c20)
  - I also have no idea how to get the mouse type to work. I accidentally got the focus to change.
  - I got the resize to work, but it took me a few tries to determine if the resize was the before or after sizes. It may be good to include them both.


3. I'm not sure I like the changes to `keybindings listen`. I kind of like what you have but I like the nerdy output before.
Before (notice it has both press and release events)
```
❯ keybindings listen
Type any key combination to see key details. Press ESC to abort.
code: Enter, modifier: NONE, flags: 0b000000
char: a, code: 0x000061, modifier: NONE, flags: 0b000000, kind: Press, state: NONE
char: a, code: 0x000061, modifier: NONE, flags: 0b000000, kind: Release, state: NONE
char: s, code: 0x000073, modifier: NONE, flags: 0b000000, kind: Press, state: NONE
char: s, code: 0x000073, modifier: NONE, flags: 0b000000, kind: Release, state: NONE
char: d, code: 0x000064, modifier: NONE, flags: 0b000000, kind: Press, state: NONE
char: d, code: 0x000064, modifier: NONE, flags: 0b000000, kind: Release, state: NONE
char: f, code: 0x000066, modifier: NONE, flags: 0b000000, kind: Press, state: NONE
char: f, code: 0x000066, modifier: NONE, flags: 0b000000, kind: Release, state: NONE
code: Enter, modifier: NONE, flags: 0b000000
code: Enter, modifier: NONE, flags: 0b000000
```
After
```
❯ keybindings listen
Type any key combination to see key details. Press ESC to abort.
type: key, key: {type: char key: a}, modifiers: []
type: key, key: {type: char key: s}, modifiers: []
type: key, key: {type: char key: d}, modifiers: []
type: key, key: {type: char key: f}, modifiers: []
type: key, key: {type: other key: enter}, modifiers: []
```
I think it would be good to keep the information we had before somehow. I also find this new format harder to understand. I wonder if we can make it a nushell table instead. Either that or how it printed previously would be better imo.

fdncred commented:
@NotLebedev Your thoughts on the input here?

NotLebedev commented:
Considering naming of command. `input listen` is fine. I have no opinion here.

Considering conversion to table. I don't understand what format should be used. I picked the format it is now, becase
1. Its close to crossterm output
2. Its easy to use with `match` command
The first point is for ease of development and is irrelevant. However the second one is more important IMO. Once again I don't quite understand how to format this into table, an example would be helpful. 

Considering `keybindings listen`, I agree. I will rollback it and add std command that utilizes new command in a loop. Hopefully performance is not too bad.

NotLebedev commented:
> In regards to record vs table, I think I'd rather just not have nested items, if possible. So, we could have `type`, `key_type`, `key_char`, `modifiers`, but maybe modifiers has to be a list because the only other way around it is to have a column/key for each modifier which probably wouldn't be good. we may also want to add `key_code`, `flags`, and `kind` for completeness. Thoughts?

I see. I misunderstood you. No problem, i totally can flatten `key` subrecord. In fact all other types of events are already flat, just an oversight on my part. Like `{ type: key key_type: <key_type> char: <string> modifiers: [ <modifier> ... ] }`.

Raw key code and flags might be useful indeed. Maybe add `--verbose` flag, to add them to output when needed?

NotLebedev commented:
I addressed all the sugestions. Also added a script to std `input display` that mirrors functionality of `keybindings listen`.

fdncred commented:
Hey @NotLebedev Thanks for updating this.
I've played with all the features I think. I'm trying to reconcile the differences here between the raw keys and how they're different than what `keybindings listen` shows. I think it would be nice if the output of `input listen -r` matches the values in `keybindings listen`.
![image](https://github.com/nushell/nushell/assets/343840/7b67d334-f068-490e-8d39-2b47c0b5514f)

I'd also like to know if you've been able to produce any events other than key and resize events. I can't seem to trigger mouse, paste, or focus.

I like your script too!

NotLebedev commented:
> I'd also like to know if you've been able to produce any events other than key and resize events. I can't seem to trigger mouse, paste, or focus.

I was, I just needed to print special codes like `\e[?1000h` to get mouse e.g. However I looked a bit deeper and discovered that crossterm can enable them via `execute` macro. I think I will add code that will automatically enable this modes when `input listen` is used. This is more convinient, I just did not know that crossterm can do this.  

> I think it would be nice if the output of input listen -r matches the values in keybindings listen.

Will do this. I just need a hint, how to print `int` type as binary (for `raw_modifiers`) and hex for `raw_code`

fdncred commented:
> Will do this. I just need a hint, how to print int type as binary (for raw_modifiers) and hex for raw_code

I think you can use `format!("{:#08x}", blah)` for hex and `format!("{:#08b}", blah)` for binary.

NotLebedev commented:
> I think you can use `format!("{:#08x}", blah)` for hex and `format!("{:#08b}", blah)` for binary.

Currently `raw_code` and `raw_modifiers` are returned by `input listen` as `Value::Int` and i was thinking about using some nushell command to convert to hex in `input display` script. I can output this fields as strings but this is not as good IMO.

fdncred commented:
ok, you could tray making them nushell literals like `0b101` and `0x101` and see what happens?

fdncred commented:
ya, i think a string representation is probably best. i'm pretty sure what `keybindings listen` does.

NotLebedev commented:
Addressed almost all remaining issues and suggestion. Fixed `input display` script and changed output of raw key and modifiers via `fmt` command. Also now all events except "bracketed paste" are always captured correctly. Only bracketed paste remains. I am investigating why it is not working.

NotLebedev commented:
Seems like bracketed paste does not work at windows at all. I tried a little bit and even some software like neovim that correctly registers bracketed paste on Linux does not work for me on Windows.

On Linux, on the other hand, it works just fine.

fdncred commented:
> Seems like bracketed paste does not work at windows at all.

I think this may be true. I think @WindSoilder saw the same thing when working with it. @WindSoilder can you confirm if bracked paste works in Windows or not?

fdncred commented:
I tested this on MacOS in WezTerm and was able to get all the event to read/print. Nice work.

WindSoilder commented:
Sorry I don't have windows machine yet(may get it in 2 days), but I don't think it works on windows, here is a relative upstream issue: https://github.com/crossterm-rs/crossterm/issues/737

fdncred commented:
I think we should move forward on this. thanks for your hard work @NotLebedev!
